src/server/game/Movement/MotionMaster.h

```diff
@@ -112,33 +122,32 @@ class TC_GAME_API MotionMaster
         void MoveRotate(uint32 time, RotateDirection direction);
         void MoveFormation(uint32 id, Position destination, uint32 moveType, bool forceRun = false, bool forceOrientation = false);
 
-        void LaunchMoveSpline(Movement::MoveSplineInit&& init, uint32 id = 0, MovementSlot slot = MOTION_SLOT_ACTIVE, MovementGeneratorType type = EFFECT_MOTION_TYPE);
+        void LaunchMoveSpline(Movement::MoveSplineInit&& init, uint32 id = 0, MovementSlots slot = MOTION_SLOT_ACTIVE, MovementGeneratorType type = EFFECT_MOTION_TYPE);
     private:
         typedef std::vector<MovementGenerator*> MovementList;
 
-        void pop();
-
-        bool NeedInitTop() const;
-        void InitTop();
-
-        void Mutate(MovementGenerator* m, MovementSlot slot);
+        void AddCleanFlag(uint8 const flag) { _cleanFlag |= flag; }
+        bool HasCleanFlag(uint8 const flag) const { return (_cleanFlag & flag) != 0; }
+        void RemoveCleanFlag(uint8 const flag) { _cleanFlag &= ~flag; }
 
+        MovementGenerator* pop();
+        void Mutate(MovementGenerator* movement, MovementSlots slot);
         void DirectClean(bool reset);
         void DelayedClean();
-        void DirectClean(MovementSlot slot);
-        void DelayedClean(MovementSlot slot);
+        void DirectClean(MovementSlots slot);
+        void DelayedClean(MovementSlots slot);
         void DirectExpire(bool reset);
         void DelayedExpire();
-        void DirectDelete(MovementGenerator* curr);
-        void DelayedDelete(MovementGenerator* curr);
+        void DirectDelete(MovementGenerator* movement);
+        void DelayedDelete(MovementGenerator* movement);
         void ClearExpireList();
 
-        MovementGenerator* _slot[MAX_MOTION_SLOT];
-        bool _initialize[MAX_MOTION_SLOT];
-        MovementList _expireList;
         Unit* _owner;
-        int _top;
+        MovementSlot* _slots[MAX_MOTION_SLOT];
+        MovementList _expireList;
+        uint8 _top;
         uint8 _cleanFlag;
+        std::unordered_multimap<uint32, MovementGenerator*>  _baseUnitStatesMap;
```

Comment:
the idea is to have an exact control of the unitstates that the owner must have and must remove.

example:
 * moment 1
SLOT 2 -> PointMovement -> ROAMING State
SLOT 3 -> Empty
* moment 2
SLOT 2 -> PointMovement -> ROAMING State
SLOT 3 -> new Generator -> PointMovement -> ROAMING State
* moment 3
SLOT 2 -> want to erase -> PointMovement -> Finalize -> wants to remove ROAMING State
SLOT 3 -> still active PointMovement -> still wants ROAMING State